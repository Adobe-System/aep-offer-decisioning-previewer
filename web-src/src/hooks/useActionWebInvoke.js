import { useState, useEffect } from "react";
import actionWebInvoke from "../utils";

export const useActionWebInvoke = ({
  actionName,
  headers = {},
  params = {},
}) => {
  const [data, setData] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    // invoke Action
    actionWebInvoke(actionName, headers, params)
      .then((response) => {
        setData(response);
        setIsLoading(false);
      })
      .catch((e) => {
        setData(null);
        setIsLoading(false);
        setError(e);
      });
  }, []);
  return { data, isLoading, error };
};
