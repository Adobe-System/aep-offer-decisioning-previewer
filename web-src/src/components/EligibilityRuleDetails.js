import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import {
  Content,
  View,
  ProgressCircle,
  IllustratedMessage,
  Heading,
  Well,
} from "@adobe/react-spectrum";
import { useActionWebInvoke } from "../hooks/useActionWebInvoke";
import {
  ProfileProvider,
  useProfileState,
  useProfileDispatch,
} from "../context/ProfileViewContext.js";
import Error from "@spectrum-icons/illustrations/Error";

const EligibilityRuleDetails = (props) => {
  const setProfileAttributes = useProfileDispatch();

  let headers = {};
  if (props.ims.token && !headers.authorization) {
    headers.authorization = `Bearer ${props.ims.token}`;
  }
  if (props.ims.org && !headers["x-gw-ims-org-id"]) {
    headers["x-gw-ims-org-id"] = props.ims.org;
  }

  const ruleDetails = useActionWebInvoke({
    actionName: "get-decision-rules",
    headers: headers,
    params: {
      containerID: props.containerID,
      ruleID: props.ruleID,
    },
  });

  let content = (
    <ProgressCircle
      id="offer-details-progress-circle"
      aria-label="Getting Offer Details"
      isIndeterminate
      isHidden={!ruleDetails.isLoading}
      marginStart="size-100"
    />
  );

  if (!ruleDetails.isLoading && ruleDetails.error) {
    content = (
      <IllustratedMessage>
        <Error />
        <Heading>Error 500: Internal server error</Heading>
        <Content>Something went wrong. Please try again later.</Content>
        <Content>{ruleDetails.error.message}</Content>
      </IllustratedMessage>
    );
  }
  if (!ruleDetails.data && !ruleDetails.error && !ruleDetails.isLoading) {
    content = (
      <View>
        <Well role="ruleDetails">
          <h3>Decision Rule Details</h3>
          <p>
            <strong>No Details found.</strong>
          </p>
        </Well>
      </View>
    );
  }
  if (!ruleDetails.isLoading && ruleDetails.data) {
    const detailedResult = ruleDetails.data._embedded.results[0];
    const ruleName = detailedResult._instance["xdm:name"];

    const ruleConditionValue =
      detailedResult._instance["xdm:condition"]["xdm:value"];

    content = (
      <View>
        <Well role="ruleDetails">
          <h3>Decision Rule Details</h3>
          <p>
            <strong>Name: </strong>
            {ruleName}
            <br /> <strong>Condition: </strong>
            {ruleConditionValue}
          </p>
        </Well>
      </View>
    );

    setProfileAttributes(ruleConditionValue );
  }

  return content;
};

EligibilityRuleDetails.propTypes = {
  offer: PropTypes.any,
};

export default EligibilityRuleDetails;
