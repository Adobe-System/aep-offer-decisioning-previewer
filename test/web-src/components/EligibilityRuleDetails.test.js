import React from "react";
import { Provider, defaultTheme } from "@adobe/react-spectrum";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
jest.mock("../../../web-src/src/hooks/useActionWebInvoke.js");
import { useActionWebInvoke } from "../../../web-src/src/hooks/useActionWebInvoke";
import mock from "../../actions/mock";
import {
  ProfileProvider,
  useProfileState,
  useProfileDispatch,
} from "../../../web-src/src/context/ProfileViewContext";
import EligibilityRuleDetails from "../../../web-src/src/components/EligibilityRuleDetails";

afterEach(cleanup);
beforeEach(() => {
  jest.clearAllMocks();
});

const onSelectionChange = jest.fn(0);
const fakeProps = {
  ims: { token: "fake-token", org: "fake-org" },
  containerID: "fake-containerID",
  ruleID: "fake-rule-id",
};

describe("<EligibilityRuleDetails> calls custom hook", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { token: "fake-token", org: "fake-org" },
      containerID: "fake-containerID",
      ruleID: "fake-rule-id",
    };

    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(
      <ProfileProvider>
        <EligibilityRuleDetails {...fakeProps} />
      </ProfileProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { authorization: "fake-token", "x-gw-ims-org-id": "fake-org" },
      onSelectionChange: onSelectionChange,
    };

    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(
      <ProfileProvider>
        <EligibilityRuleDetails {...fakeProps} />{" "}
      </ProfileProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("<EligibilityRuleDetails> on load", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("renders correctly on load", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(
      <ProfileProvider>
        <EligibilityRuleDetails {...fakeProps} />
      </ProfileProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("has a <ProgressCircle> when loading", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(
      <ProfileProvider>
        <EligibilityRuleDetails {...fakeProps} />
      </ProfileProvider>
    );
    expect(screen.getByLabelText("Getting Offer Details")).toBeDefined();
  });
});

describe("<EligibilityRuleDetails> on load", () => {
  const mockResponse = {
    data: mock.data.offerDecisionRule,
    isLoading: false,
    error: null,
  };
  it("renders correctly on data", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(
      <Provider theme={defaultTheme} colorScheme={`light`}>
        <ProfileProvider>
          <EligibilityRuleDetails {...fakeProps} />
        </ProfileProvider>
      </Provider>
    );
    expect(screen.getByText("Decision Rule Details")).toBeInTheDocument();
    expect(screen.getByText("Condition:")).toBeInTheDocument();
  });
});

describe("<EligibilityRuleDetails> on error", () => {
  const mockResponse = {
    data: null,
    isLoading: false,
    error: { message: "fake-error-message" },
  };
  it("renders correctly on error", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(
      <ProfileProvider>
        <EligibilityRuleDetails {...fakeProps} />
      </ProfileProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("show error message", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(
      <ProfileProvider>
        <EligibilityRuleDetails {...fakeProps} />
      </ProfileProvider>
    );
    expect(screen.getAllByText("fake-error-message")).toBeDefined();
  });
});

describe("<EligibilityRuleDetails> on data", () => {
  const mockResponse = {
    data: null,
    isLoading: false,
    error: null,
  };
  it("renders correctly on no data", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(
      <ProfileProvider>
        <EligibilityRuleDetails {...fakeProps} />
      </ProfileProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("show no data message", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(
      <ProfileProvider>
        <EligibilityRuleDetails {...fakeProps} />
      </ProfileProvider>
    );
    expect(screen.getAllByText("No Details found.")).toBeDefined();
  });
});
