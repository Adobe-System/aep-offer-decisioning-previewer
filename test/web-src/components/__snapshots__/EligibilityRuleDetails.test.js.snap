// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EligibilityRuleDetails> calls custom hook call custom hooks with global parameters correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Details"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="offer-details-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<EligibilityRuleDetails> calls custom hook call custom hooks with global parameters correctly 2`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Details"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="offer-details-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<EligibilityRuleDetails> on data renders correctly on no data 1`] = `
<DocumentFragment>
  <div>
    <div
      class="_spectrum-Well_daead"
      role="ruleDetails"
    >
      <h3>
        Decision Rule Details
      </h3>
      <p>
        <strong>
          No Details found.
        </strong>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<EligibilityRuleDetails> on error renders correctly on error 1`] = `
<DocumentFragment>
  <div
    class="_flex_ffbb7 _spectrum-IllustratedMessage_5a24b"
  >
    <svg
      aria-hidden="true"
      height="94"
      width="146.569"
    >
      <g
        transform="translate(-392.731 -2966.5)"
      >
        <path
          d="M415.8 2988v-18a2.006 2.006 0 012-2h118a2.006 2.006 0 012 2v23a2.006 2.006 0 01-2 2h-110M428.8 3000h107a2.006 2.006 0 012 2v23a2.006 2.006 0 01-2 2h-92M443.8 3032h92a2.006 2.006 0 012 2v23a2.006 2.006 0 01-2 2h-118a2.006 2.006 0 01-2-2v-20M526.8 3028v3M526.8 2996v3"
          fill="none"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-miterlimit="10"
          stroke-width="3"
        />
        <circle
          cx="4"
          cy="4"
          fill="none"
          r="4"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-miterlimit="10"
          stroke-width="2"
          transform="translate(522.8 2977.5)"
        />
        <circle
          cx="4"
          cy="4"
          fill="none"
          r="4"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-miterlimit="10"
          stroke-width="2"
          transform="translate(522.8 3010)"
        />
        <circle
          cx="4"
          cy="4"
          fill="none"
          r="4"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-miterlimit="10"
          stroke-width="2"
          transform="translate(522.8 3041.5)"
        />
        <path
          d="M413.1 2995.5l-18.4 31.1a2.916 2.916 0 002.6 4.4h36.9a2.933 2.933 0 002.6-4.4l-18.4-31.1a3.091 3.091 0 00-5.3 0zM415.8 3016.2v-6.9"
          fill="none"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-miterlimit="10"
          stroke-width="3"
        />
        <circle
          cx="1.9"
          cy="1.9"
          r="1.9"
          stroke="none"
          transform="translate(413.9 3020.3)"
        />
      </g>
    </svg>
    <h3
      class="_spectrum-IllustratedMessage-heading_5a24b"
    >
      Error 500: Internal server error
    </h3>
    <section
      class="_spectrum-IllustratedMessage-description_5a24b"
    >
      Something went wrong. Please try again later.
    </section>
    <section
      class="_spectrum-IllustratedMessage-description_5a24b"
    >
      fake-error-message
    </section>
  </div>
</DocumentFragment>
`;

exports[`<EligibilityRuleDetails> on load renders correctly on load 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Details"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="offer-details-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
